landa/water_body_management/change_log.py

```diff
@@ -0,0 +1,200 @@
+				"datetime": entry.creation,
+			}
+
+		# Modified Water Body/Fish Species Log
+		change_log = self._build_modified_change_log(entry, changed_data, event)
+		if change_log:
+			return change_log
+
+		return None
+
+	def _get_version_log_query(self, from_datetime: str):
+		version = frappe.qb.DocType("Version")
+		file = frappe.qb.DocType("File")
+		return (
+			frappe.qb.from_(version)
+			.left_join(file)
+			.on((version.ref_doctype == "File") & (version.docname == file.name))
+			.select(
+				version.name,
+				version.ref_doctype.as_("doctype"),
+				version.docname,
+				version.creation,
+				version.data,
+				ConstantColumn(0).as_("deleted"),
+				file.attached_to_doctype,
+				file.attached_to_name,
+			)
+			.where(
+				version.creation >= from_datetime,
+			)
+			.where(
+				version.ref_doctype.isin(
+					[
+						"Water Body",
+						"Fish Species",
+						"File",
+						"Water Body Management Local Organization",
+						"Water Body Rules",
+					]
+				)
+			)
+			.where((file.attached_to_doctype.isnull()) | (file.attached_to_doctype == "Water Body"))
+		)
+
+	def _get_deleted_document_query(self, from_datetime: str):
+		# Deleted Document DocType is not joined with File DocType
+		# because the file does not exist anymore
+		deleted_document = frappe.qb.DocType("Deleted Document")
+		return (
+			frappe.qb.from_(deleted_document)
+			.select(
+				deleted_document.name,
+				deleted_document.deleted_doctype.as_("doctype"),
+				deleted_document.deleted_name.as_("docname"),
+				deleted_document.creation,
+				deleted_document.data,
+				ConstantColumn(1).as_("deleted"),
+				ConstantColumn("").as_("attached_to_doctype"),
+				ConstantColumn("").as_("attached_to_name"),
+			)
+			.where(
+				deleted_document.creation >= from_datetime,
+			)
+			.where(
+				deleted_document.deleted_doctype.isin(
+					[
+						"Water Body",
+						"Fish Species",
+						"File",
+						"Water Body Management Local Organization",
+					]
+				)
+			)
+		)
+
+	def _get_event(self, entry, changed_data):
+		"""Return the event of the document (Created, Modified, Deleted)."""
+		if cint(entry.deleted):
+			return "Deleted"
+
+		return "Created" if changed_data.updater_reference else "Modified"
+
+	def _build_modified_change_log(self, entry, changed_data, event):
+		"""Return a pretty dict with changes for modified Water Body/Fish Species."""
+		change_log = {
+			"id": entry.name,
+			"doctype": entry.doctype,
+			"docname": entry.docname,
+			"datetime": entry.creation,
+			"event": event,
+			"changes": {},
+		}
+
+		if not changed_data:
+			return None
+
+		# Table fields changes
+		for key in ["added", "removed", "row_changed"]:
+			change_log["changes"].update(
+				{row[0]: None for row in changed_data.get(key) if row[0] not in change_log["changes"]}
+			)
+
+		# Other fields changes
+		for row in changed_data.get("changed"):
+			key = "geojson" if row[0] == "location" else row[0]
+			change_log["changes"].update({key: row[2]})
```

Comment:
```suggestion
		for row in changed_data.get("changed"):
			if row[0] == "location":
				key = "geojson"
				data = json.loads(row[2]) if row[2] else None
			else:
				key = row[0]
				data = row[2]

			change_log["changes"][key] = data
```

Let's avoid stringified JSON as a value in a  bigger JSON structure, where possible. This way we can still stringify the bigger structure without getting into stringification hell.

Or, to keep the change logs small and simple (we originally planned to support simple data types only):

```suggestion
		for row in changed_data.get("changed"):
			key, data = ("geojson", None) if row[0] == "location" else (row[0], row[2])
			change_log["changes"][key] = data
```